Fixed PEP8 in Pong examples

erm3nda opened the pull request:
In short:

- fixed pep8 in previous step files.
- Added a bounce controller as step6
- Added step6 to docs as "Improved bouncing"

I found that when you hit the ball with upper or lower side of the paddle, the bounce gets called many many times, creating a undesired behavior. Enabling/disabling bounce ability like this solves the problem.

I hope this time pass checks.

tshirtman commented:
Reading the code, i'm thinking there might be an easier solution than using a timer.

You could keep the "can_bounce" property, but instead of re-enabling it by timer, just check its value in `bounce_ball` and if it's False *and* it doesn't collide, then you can set it to True again. The `can_bounce` property could also be moved to either the paddle or the ball, depending on what's more convenient, and avoid the need of the `PongBouncer` class.

Would that solution work for you?

erm3nda commented:
@tshirtman My first try was to add a global can_bounce and another if with 
    
    if not self.collide_widget(ball) and not self.parent.can_bounce: 
        self.parent.can_bounce = 1

But didn't work for some reason. I don't know how is all data handled internally. Feel free to try yourself. Anyway, there's another better way to fix that without use timer and the callback.

erm3nda commented on doc/sources/tutorials/pong.rst:
```diff
@@ -431,6 +431,90 @@ pong.kv:
+        
+Thats's all, now the ball will bounce just once per collision.
+
+Limiting ball speed to 26. To avoid current game round ending due to ball speed 
```
I found a cleaner way to do that without timer, using `width` and `x` values from widgets.
I've installed autoPEP8 plugin but seems is not enough.

I'll install some more plugins and try again.

Thanks.

erm3nda commented:
Doing a clean PR again :-(

tshirtman commented:
You know you don't need to close these PR every time, you can just create reset your branch to master, do your new solution, commit and push --force to this branch, that will update the PR.

KeyWeeUsr commented:
@erm3nda I just randomly found some useful articles regarding managing the branches, they'll help you:

* https://www.atlassian.com/git/tutorials/merging-vs-rebasing
* http://gitforteams.com/resources/rebasing.html
* https://www.youtube.com/watch?v=u-kAeG4jkMA

Unless you're using a public branch (branch used by more people than only you), you can happily do `git rebase`, `git reset` and what not however you like. In the end if we don't agree with how the commits/messages are structured we can do either `Squash and merge`through github interface or just merge via console and pick specific commits + squash the less pretty ones. You should feel comfortable doing `rebase`, `merge` and `pull`/`push` changes in your own branch (such as this one for this particular PR) :)

erm3nda commented:
It is possible to remove some commits from this PR and merge it?
I don't want to be repeating myself many more times. Please. keep those commits that are not about the bouncer stuff, just fixing old and current errors.

Pong tutorial: fixed multi-line imports  …
a456baf

Pong tutorial: fixed PEP8  …
7664780

Pong tutorial: added Buildozer mention in docs  …
2114995

I'll try the other way to do that wihout adding a new Class and without using time this time and creating a PR for it separately. Last time, im sorry. It's being time consuming for something that is not that much important (fixing a little tut).

Regards.

tshirtman commented:
will look into it, but you could really achieve the same effect by doing a git rebase.

about the way to test, this is what i was thinking about, is it what you tried?

```diff
diff --git a/examples/tutorials/pong/main.py b/examples/tutorials/pong/main.py
index c007a2406..009fcfced 100644
--- a/examples/tutorials/pong/main.py
+++ b/examples/tutorials/pong/main.py
@@ -4,21 +4,25 @@ kivy.require('1.1.1')
 from kivy.app import App
 from kivy.uix.widget import Widget
 from kivy.properties import NumericProperty, ReferenceListProperty,\
-    ObjectProperty
+    ObjectProperty, BooleanProperty
 from kivy.vector import Vector
 from kivy.clock import Clock
 
 
 class PongPaddle(Widget):
     score = NumericProperty(0)
+    can_collide = BooleanProperty(True)
 
     def bounce_ball(self, ball):
-        if self.collide_widget(ball):
+        if self.can_collide and self.collide_widget(ball):
             vx, vy = ball.velocity
             offset = (ball.center_y - self.center_y) / (self.height / 2)
             bounced = Vector(-1 * vx, vy)
             vel = bounced * 1.1
             ball.velocity = vel.x, vel.y + offset
+            self.can_collide = False
+        elif not self.can_collide and not self.collide_widget(ball):
+            self.can_collide = True
 
 
 class PongBall(Widget):
```

I disagree about it being not very important, it's tutorial code, it's something newcomers will look at as good practice, and take inspiration from (and for more experienced coders, might make them reluctant to use the framework if the code seems too complicated for a simple task).

If this solution doesn't work, and the time based one does, we can merge it, but i'd like to understand why this one doesn't first.

erm3nda commented:
Trying to use this PR to fix old code PEP8 and deferring the bounce thing untill i found a nicer solution.

erm3nda commented:
Well i think i have it, i've reset to those 2 commits.
I'll fix the bounce thing in another PR.

Regards.

erm3nda commented:
> "Build execution time has reached the maximum allowed time for your plan (60 minutes).

C'mon this is becoming a total waste of time. What is supposed I should to now? ...

erm3nda commented:
Hi @tshirtman 

I tried today in a new and clean install of Kivy over Win10. This time it worked as expected, the ball bounced just once.

`class PongPaddle(Widget):
     score = NumericProperty(0)
+    can_collide = BooleanProperty(True)
 
     def bounce_ball(self, ball):
-        if self.collide_widget(ball):
+        if self.can_collide and self.collide_widget(ball):
             vx, vy = ball.velocity
             offset = (ball.center_y - self.center_y) / (self.height / 2)
             bounced = Vector(-1 * vx, vy)
             vel = bounced * 1.1
             ball.velocity = vel.x, vel.y + offset
+            self.can_collide = False
+        elif not self.can_collide and not self.collide_widget(ball):
+            self.can_collide = True`

I cannot understand why went wrong with my last try. I didn't nothing special this time.
I'll try again with my old scenario, but with a clean install too.

Regards.
